lib/util/extractSourceMap.js

```diff
@@ -0,0 +1,290 @@
+			if (err) return callback(err);
+			const { path: sourceURL, data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	} else {
+		// 4. Relative path
+		const sourceURL = getAbsolutePath(fs, context, url);
+		if (skipReading) return callback(null, { sourceURL });
+
+		fetchFromFilesystem(fs, sourceURL, (err, result) => {
+			if (err) return callback(err);
+			const { data: sourceContent } = result;
+			callback(null, { sourceURL, sourceContent });
+		});
+	}
+}
+
+/**
+ * @param {string} input input data
+ * @param {InputFileSystem} fs filesystem
+ * @param {NormalModule} module normal module
+ * @param {boolean} removeOnly should only remove source map comment
+ * @param {(err: Error|null, code?: string, sourceMap?: object) => void} callback callback
+ */
+module.exports = function extractSourceMap(
+	input,
+	fs,
+	module,
+	removeOnly,
+	callback
+) {
+	const { sourceMappingURL, replacementString } = getSourceMappingURL(input);
+
+	if (!sourceMappingURL) {
+		callback(null, input);
+		return;
+	} else if (removeOnly) {
+		callback(null, input.replace(replacementString, ""));
+		return;
+	}
+
+	fetchFromURL(
+		fs,
+		module.context || "",
+		sourceMappingURL,
+		false,
+		(err, result) => {
+			if (err) return callback(err);
+			const { sourceURL, sourceContent } = result;
+
+			if (sourceURL && module.buildInfo && module.buildInfo.fileDependencies) {
+				module.buildInfo.fileDependencies.add(sourceURL);
+			}
+
+			let map;
+
+			try {
+				map = JSON.parse(sourceContent.replace(/^\)\]\}'/, ""));
+			} catch (parseError) {
+				return callback(
+					new WebpackError(
+						`Failed to parse source map from '${sourceMappingURL}': ${parseError}`
+					)
+				);
+			}
+
+			if (map.sources && map.sources.length > 0) {
+				const promises = [];
+				for (let i = 0; i < map.sources.length; i++) {
+					const k = i;
+					promises.push(
+						new Promise(resolve => {
+							const skipReading =
+								map.sourcesContent[k] !== null &&
+								map.sourcesContent[k] !== undefined;
+							fetchFromURL(
+								fs,
+								module.context || "",
+								map.sources[k],
+								skipReading,
+								(err, result) => {
+									if (err) {
+										return resolve(map.sourcesContent[k] || null);
+									}
+
+									const { sourceURL, sourceContent } = result;
+
+									if (
+										sourceURL &&
+										module.buildInfo &&
+										module.buildInfo.fileDependencies
+									) {
+										module.buildInfo.fileDependencies.add(sourceURL);
+									}
+
+									resolve(map.sourcesContent[k] || sourceContent || null);
+								}
+							);
+						})
+					);
+				}
+				if (promises.length) {
+					return Promise.all(promises).then(sourcesContent => {
+						map.sourcesContent = sourcesContent;
+						callback(null, input.replace(replacementString, ""), map);
+					});
+				}
+			}
+
+			return callback(null, input.replace(replacementString, ""), map);
+		}
+	);
+};
```

Comment:
Also let's implement https://github.com/webpack-contrib/source-map-loader/tree/master?tab=readme-ov-file#filtersourcemappingurl, sometimes some source maps are really broken and you can't do anything with them, so let's add:
```
type ExtractSourceMap = boolean | {  errorSeverity: "...", test: Rules, include: Rules, Exlcude: Rules }
```